### PR TITLE
FIO-7733: type the form component

### PR DESCRIPTION
**Pull request by brendanbond**
## Link to Jira Ticket

https://formio.atlassian.net/browse/FIO-7733

### General Updates
1. Updates the build system to use tsc rather than babel
2. Removes unused dependencies
3. Removes test runner dependencies (updated tests to follow!)
4. Updates eslint ergonomics to ensure React best practices
5. Pulls EventEmitter from @formio/js
6. Updates documentation
7. Removes stale types
8. Removes stale tests
9. Adds jest and [@testing-library/react](https://testing-library.com/docs/react-testing-library/intro/) so we can write and maintain a modern test suite

### Updates to `<Form />`
1. Fully types the component and its props
2. Reorgs the code so as not to instantiate utility functions in the render function
3. Updates the `src` prop to resemble the `createForm` API (i.e. it takes either a form definition or a form url)
4. Renames some props to conform to idiomatic prop names (see below) 
5. Consolidates effect functions
6. Updates the render element ref to use `useRef`

## Breaking Changes / Backwards Compatibility

Generally speaking the goal is to be backwards compatible here (see e.g. the `getEffectiveProps` function), but I would point the reviewer specifically towards the EventEmitter change and broadly towards testing this PR in your own existing application so we can be sure that everything works properly.

## Dependencies

* [formio.js#5516](https://github.com/formio/formio.js/pull/5516)
* [formio.js#5513](https://github.com/formio/formio.js/pull/5513)
* [formio.js#5512](https://github.com/formio/formio.js/pull/5512)
* [formio.js#5507](https://github.com/formio/formio.js/pull/5507)
* [formio.js#5506](https://github.com/formio/formio.js/pull/5506)

## How has this PR been tested?

I've loaded this PR into a simple Next.js application to ensure everything works as expected. Also, now that we've moved away from karma, I expect to add more tests in the next major PR to this library.

## Checklist:

- [x] I have completed the above PR template
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (if applicable)
- [x] My changes generate no new warnings
- [ ] My changes include tests that prove my fix is effective (or that my feature works as intended)
- [x] New and existing unit/integration tests pass locally with my changes
- [x] Any dependent changes have corresponding PRs that are listed above

### CODE REVIEW

File: src/components/Form.tsx [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+					instance.form = formSource;
+					if (url) {
+						instance.url = url;
+					}
+					if (formReadyCallback) formReadyCallback(instance);
+					setFormInstance(instance);
+				}
+			});
+		}
+	}, [formConstructor, formSource, props, url, formReadyCallback]);
```
Comment (gyanendrasinghpanwar) [inline src/components/Form.tsx:254]: @brendanbond  
I presume we have added props to take care of any props change.

I believe it also include handlers aka onAnyEvent functions(onSubmit etc..), which does not require recreating form instance.
Recreating the form instance for every props changes(specifically handlers) can lead to unnecessary performance overhead.
We only need to register them in form instance and turn off previous events on handlers change.

Ideal would be to add specific prop and create another effect for handler.

Let me know if I need to add snippet and testing for it.

Comment (brendanbond) [reply src/components/Form.tsx:254]: Yeah that's a good point - I'll make an update

**Review by gyanendrasinghpanwar — APPROVED** [review]

**Comment by gyanendrasinghpanwar** [comment]
@brendanbond Do we need PR to type the **FormBuilder** ? I do have experience on it.

**Comment by brendanbond** [comment]
@gyanendrasinghpanwar I started that today, perhaps since i'm going from top to bottom you could go bottom to top and meet in the middle, maybe you could start typing FormGrid?

**Comment by gyanendrasinghpanwar** [comment]
@brendanbond Absolutely, sounds like a solid plan! I'm ready to collaborate on getting the builder ready. Going from bottom to top works for me. Once we've got that sorted, we can dive into FormGrid or FormEdit. Feel free to share any suggestions or preferences along the way.

File: src/components/Form.tsx [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+					instance.form = formSource;
+					if (url) {
+						instance.url = url;
+					}
+					if (formReadyCallback) formReadyCallback(instance);
+					setFormInstance(instance);
+				}
+			});
+		}
+	}, [formConstructor, formSource, url, formReadyCallback, options]);
```
Comment (gyanendrasinghpanwar) [inline src/components/Form.tsx:249]:
```
useEffect(() => {
    const createInstance = async () => {
        
        if (renderElement.current === null) {
            console.warn('Form element not found');
            return;
        }

        if (typeof formSource === 'object') {
            // If formSource is an object and hasn't changed, return early
            if (isEqual(formSource, prevFormDefinition.current)) {
                return;
            }
            
            prevFormDefinition.current = cloneDeep(formSource);
        }

        const instance = await createWebformInstance(formConstructor, formSource, renderElement.current, options);

        if (instance) {
            
            if (typeof formSource === 'string') {
                instance.src = formSource;
            } else if (typeof formSource === 'object') {
                instance.form = formSource;
                
                if (url) {
                    instance.url = url;
                }
            }
            
            if (formReadyCallback) {
                formReadyCallback(instance);
            }
            
            setFormInstance(instance);
        }
    };

    createInstance();

}, [formConstructor, formSource, url, formReadyCallback, options]);

```

@brendanbond pls check if it looks more readable.

Comment (brendanbond) [reply src/components/Form.tsx:249]: @gyanendrasinghpanwar yeah, I toyed with something like this originally, but for whatever reason I really wanted to move the `createInstance` method out of the render function scope. I do like this a little better though for the lack of duplication - I'll make the change

Comment (brendanbond) [reply src/components/Form.tsx]: Updated!

**Pull request closed without merging**